In [1]:
import numpy as np
import matplotlib.pyplot as plt
from sklearn.metrics import accuracy_score, classification_report

import keras
import tensorflow as tf
from tensorflow import keras
from keras.models import Sequential
from tensorflow.keras import regularizers, optimizers
from tensorflow.keras.layers import Input, Conv2D, Dense, Flatten, Dropout
from tensorflow.keras.layers import GlobalMaxPooling2D, MaxPooling2D
from tensorflow.keras.layers import BatchNormalization
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Model

# Loading CIFAR10 DataSet for CNN Study

In [2]:
from tensorflow.keras.datasets import cifar10
(X_train, Y_train), (X_test, Y_test) = cifar10.load_data()

170498071/170498071 ━━━━━━━━━━━━━━━━━━━━ 2s 0us/step


In [3]:
X_train = X_train/255
X_test = X_test/255
Y_train_en = to_categorical(Y_train,10)
Y_test_en = to_categorical(Y_test,10)

# 3 Block CNN Study

In [4]:
model = Sequential()
model.add(Conv2D(32,(4,4),input_shape = (32,32,3),activation='relu'))
model.add(MaxPooling2D(pool_size = (2,2)))
model.add(Conv2D(32,(4,4),input_shape = (32,32,3),activation='relu'))
model.add(MaxPooling2D(pool_size = (2,2)))
model.add(Conv2D(32,(4,4),input_shape = (32,32,3),activation='relu'))
model.add(MaxPooling2D(pool_size = (2,2)))

model.add(Flatten())
model.add(Dense(128, activation ='relu'))
model.add(Dense(10, activation ='softmax'))
model.compile(loss ='categorical_crossentropy', optimizer ='adam',
metrics =['accuracy'])

model.summary()
history = model.fit(X_train, Y_train_en, epochs = 10, verbose=1,
                    validation_data=(X_test,Y_test_en))

/usr/local/lib/python3.10/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ conv2d (Conv2D)                      │ (None, 29, 29, 32)          │           1,568 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d (MaxPooling2D)         │ (None, 14, 14, 32)          │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ conv2d_1 (Conv2D)                    │ (None, 11, 11, 32)          │          16,416 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d_1 (MaxPooling2D)       │ (None, 5, 5, 32)            │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ conv2d_2 (Conv2D)                    │ (None, 2, 2, 32)            │          16,416 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d_2 (MaxPooling2D)       │ (None, 1, 1, 32)            │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ flatten (Flatten)                    │ (None, 32)                  │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense (Dense)                        │ (None, 128)                 │           4,224 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ (None, 10)                  │           1,290 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 39,914 (155.91 KB)

 Trainable params: 39,914 (155.91 KB)

 Non-trainable params: 0 (0.00 B)

Epoch 1/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 68s 42ms/step - accuracy: 0.3132 - loss: 1.8318 - val_accuracy: 0.5151 - val_loss: 1.3419
Epoch 2/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 63s 40ms/step - accuracy: 0.5172 - loss: 1.3372 - val_accuracy: 0.5497 - val_loss: 1.2618
Epoch 3/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 82s 40ms/step - accuracy: 0.5743 - loss: 1.1963 - val_accuracy: 0.5842 - val_loss: 1.1662
Epoch 4/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 84s 41ms/step - accuracy: 0.6081 - loss: 1.1027 - val_accuracy: 0.5873 - val_loss: 1.1687
Epoch 5/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 81s 41ms/step - accuracy: 0.6371 - loss: 1.0317 - val_accuracy: 0.6256 - val_loss: 1.0631
Epoch 6/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 84s 42ms/step - accuracy: 0.6553 - loss: 0.9804 - val_accuracy: 0.6397 - val_loss: 1.0304
Epoch 7/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 79s 40ms/step - accuracy: 0.6691 - loss: 0.9407 - val_accuracy: 0.6416 - val_loss: 1.0386
Epoch 8/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 83s 41ms/step - accuracy: 0.6824 -

In [5]:
test_loss, test_acc = model.evaluate(X_test, Y_test_en, verbose=2)
print(f'\nTest Accuracy 3 Block Study: {test_acc:.3f}')

313/313 - 3s - 10ms/step - accuracy: 0.6596 - loss: 0.9881

Test Accuracy 3 Block Study: 0.660


# 2 Blocks CNN Study

In [6]:
model1 = Sequential()
model1.add(Conv2D(32,(4,4),input_shape = (32,32,3),activation='relu'))
model1.add(MaxPooling2D(pool_size = (2,2)))
model1.add(Conv2D(32,(4,4),input_shape = (32,32,3),activation='relu'))
model1.add(MaxPooling2D(pool_size = (2,2)))

model1.add(Flatten())
model1.add(Dense(128, activation ='relu'))
model1.add(Dense(10, activation ='softmax'))
model1.compile(loss ='categorical_crossentropy', optimizer ='adam',
metrics =['accuracy'])

model1.summary()
history = model1.fit(X_train, Y_train_en, epochs = 10, verbose=1,
                     validation_data=(X_test,Y_test_en))

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ conv2d_3 (Conv2D)                    │ (None, 29, 29, 32)          │           1,568 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d_3 (MaxPooling2D)       │ (None, 14, 14, 32)          │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ conv2d_4 (Conv2D)                    │ (None, 11, 11, 32)          │          16,416 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d_4 (MaxPooling2D)       │ (None, 5, 5, 32)            │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ flatten_1 (Flatten)                  │ (None, 800)                 │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_2 (Dense)                      │ (None, 128)                 │         102,528 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_3 (Dense)                      │ (None, 10)                  │           1,290 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 121,802 (475.79 KB)

 Trainable params: 121,802 (475.79 KB)

 Non-trainable params: 0 (0.00 B)

Epoch 1/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 71s 44ms/step - accuracy: 0.3812 - loss: 1.6963 - val_accuracy: 0.5346 - val_loss: 1.2978
Epoch 2/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 78s 42ms/step - accuracy: 0.5625 - loss: 1.2307 - val_accuracy: 0.5919 - val_loss: 1.1364
Epoch 3/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 83s 43ms/step - accuracy: 0.6260 - loss: 1.0706 - val_accuracy: 0.6197 - val_loss: 1.0800
Epoch 4/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 83s 43ms/step - accuracy: 0.6563 - loss: 0.9803 - val_accuracy: 0.6440 - val_loss: 1.0192
Epoch 5/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 79s 41ms/step - accuracy: 0.6848 - loss: 0.9058 - val_accuracy: 0.6553 - val_loss: 0.9969
Epoch 6/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 81s 40ms/step - accuracy: 0.7089 - loss: 0.8364 - val_accuracy: 0.6530 - val_loss: 1.0215
Epoch 7/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 67s 43ms/step - accuracy: 0.7305 - loss: 0.7706 - val_accuracy: 0.6726 - val_loss: 0.9719
Epoch 8/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 82s 43ms/step - accuracy: 0.7498 -

In [7]:
test_loss, test_acc = model1.evaluate(X_test, Y_test_en, verbose=2)
print(f'\nTest accuracy 2 Block Study: {test_acc:.3f}')

313/313 - 3s - 10ms/step - accuracy: 0.6596 - loss: 0.9881

Test accuracy 2 Block Study: 0.660


# 1 Block CNN Study

In [9]:
model2 = Sequential()
model2.add(Conv2D(32,(4,4),input_shape = (32,32,3),activation='relu'))
model2.add(MaxPooling2D(pool_size = (2,2)))

model2.add(Flatten())
model2.add(Dense(128, activation ='relu'))
model2.add(Dense(10, activation ='softmax'))
model2.compile(loss ='categorical_crossentropy', optimizer ='adam',
metrics =['accuracy'])

model2.summary()
history = model2.fit(X_train, Y_train_en, epochs = 10,
verbose=1,validation_data=(X_test,Y_test_en))

/usr/local/lib/python3.10/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ conv2d_6 (Conv2D)                    │ (None, 29, 29, 32)          │           1,568 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d_6 (MaxPooling2D)       │ (None, 14, 14, 32)          │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ flatten_3 (Flatten)                  │ (None, 6272)                │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_6 (Dense)                      │ (None, 128)                 │         802,944 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_7 (Dense)                      │ (None, 10)                  │           1,290 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 805,802 (3.07 MB)

 Trainable params: 805,802 (3.07 MB)

 Non-trainable params: 0 (0.00 B)

Epoch 1/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 57s 36ms/step - accuracy: 0.4033 - loss: 1.6532 - val_accuracy: 0.5592 - val_loss: 1.2493
Epoch 2/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 81s 35ms/step - accuracy: 0.5867 - loss: 1.1764 - val_accuracy: 0.6082 - val_loss: 1.1088
Epoch 3/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 82s 35ms/step - accuracy: 0.6439 - loss: 1.0229 - val_accuracy: 0.6273 - val_loss: 1.0552
Epoch 4/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 56s 36ms/step - accuracy: 0.6751 - loss: 0.9319 - val_accuracy: 0.6416 - val_loss: 1.0583
Epoch 5/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 82s 35ms/step - accuracy: 0.7084 - loss: 0.8370 - val_accuracy: 0.6358 - val_loss: 1.0507
Epoch 6/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 82s 36ms/step - accuracy: 0.7390 - loss: 0.7568 - val_accuracy: 0.6433 - val_loss: 1.0747
Epoch 7/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 82s 36ms/step - accuracy: 0.7560 - loss: 0.7044 - val_accuracy: 0.6329 - val_loss: 1.0941
Epoch 8/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 82s 35ms/step - accuracy: 0.7787 -

In [10]:
test_loss, test_acc = model2.evaluate(X_test, Y_test_en, verbose=2)
print(f'\nTest accuracy for 1 Block Study: {test_acc:.3f}')

313/313 - 2s - 8ms/step - accuracy: 0.6463 - loss: 1.1989

Test accuracy for 1 Block Study: 0.646


# Conclusion (CIFAR 10 CNN Study):

If your problem requires detailed feature extraction, decreasing the number of blocks too much may decrease the performance.

In 2 Block and 3 Block CNN Study, accuracy (0.66) didn't changed but in 1 Block CNN, it got decreased to (0.646).

CIFAR10 shows lesser accuracy than FashionMNIST at the same no. of hidden layers, which means CIFAR10 is more complex dataset than FashionMNIST. CIFAR 10 requires more no. of hidden layers to reach the accuracy level of that of the FashionMNIST. 